mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp

```diff
@@ -2079,18 +2079,19 @@ LogicalResult transform::detail::verifyTransformOpInterface(Operation *op) {
 // Entry point.
 //===----------------------------------------------------------------------===//
 
-LogicalResult
-transform::applyTransforms(Operation *payloadRoot,
-                           TransformOpInterface transform,
-                           const RaggedArray<MappedValue> &extraMapping,
-                           const TransformOptions &options) {
+LogicalResult transform::applyTransforms(
+    Operation *payloadRoot, TransformOpInterface transform,
+    const RaggedArray<MappedValue> &extraMapping,
+    const TransformOptions &options, bool enforceToplevelTransformOp) {
 #ifndef NDEBUG
-  if (!transform->hasTrait<PossibleTopLevelTransformOpTrait>() ||
-      transform->getNumOperands() != 0) {
-    transform->emitError()
-        << "expected transform to start at the top-level transform op";
-    llvm::report_fatal_error("could not run transforms",
-                             /*gen_crash_diag=*/false);
+  if (enforceToplevelTransformOp) {
```

Comment:
I suspect we may need a check that whatever is the `transform` operation, it has a region with the number of arguments (and their types) matching `payloadRoot` + `extraMapping` when the trait is not enforced.

Reply:
Calling `detail::verifyPossibleTopLevelTransformOpTrait` directly provides part of this in the absence of an op actually subscribing to `PossibleTopLevelTransformOpTrait`.

The extra mappings part is only used in the TransformState.

In the future, we may want to reconsider the top-level trait altogether.